File: batou-rules/testdata/fixtures/go/safe/header_injection_safe.go
package handler

import (
	"net/http"
	"net/url"
	"strings"
)

func setCustomHeaderSafe(w http.ResponseWriter, r *http.Request) {
	// SAFE: URL-encode the value before setting header
	customVal := r.URL.Query().Get("val")
	sanitized := url.QueryEscape(customVal)
	w.Header().Set("X-Custom", sanitized)

	// SAFE: strip CRLF before setting header
	user := r.FormValue("user")
	user = strings.ReplaceAll(user, "\r", "")
	user = strings.ReplaceAll(user, "\n", "")
	w.Header().Set("X-Forwarded-User", user)
}

File: batou-rules/testdata/fixtures/javascript/safe/header_injection_safe.ts
import express from 'express';

const app = express();

app.get('/api/data', (req, res) => {
  // SAFE: encode value before setting header
  const requestId = encodeURIComponent(req.query.requestId || '');
  res.setHeader('X-Request-Id', requestId);
  res.json({ ok: true });
});
